Pluggable auth

ChristophWurst opened the pull request:
~~This revives https://github.com/owncloud/core/pull/1975~~

I removed the two-factor code from this PR to keep it simple.

TODO:
- [x] Create session token for new browser session (login)
  - [x] Save hashed token + encrypted login password
- [x] Invalidate (delete) session token on logout
- [x] Add background job to kill old session tokens
  - [x] Don't kill permanent client tokens
- [x] Check browser session on every request -> logout if invalidated/not found
- [x] Check login credentials each x minutes and log the user out if the check fails
- [x] Update token 'last_activity' in every request
- [x] ~~Invalidate tokens if the session is killed [in base.php](https://github.com/owncloud/core/blob/14c34919774484d095d26ad2a7246fc897dc2d41/lib/base.php#L433-L439)?~~ -> #24542
- [x] Add controller where clients can get a new access token
  - Can be tested with `curl --request POST -F 'user=youruser' -F 'password=yourpassword' localhost:8080/token/generate`
- [x] ~~Don't use sessions if the client sends an auth token~~ -> #24543
- [x] Add DB index on 'last_activity'
- [x] Add token type column
- [x] Migrate [OCS auth](https://github.com/owncloud/core/blob/887045b790ced3473f2c0e955abcdd2bc1944612/lib/private/api.php#L330-L384)
- [x] Adjust sabre auth backend
- [x] Increase version number to trigger db migration
- [x] Fix existing unit tests
- [x] Add new unit tests for token auth
- [x] Fix existing integration test
- [x] Fix apache auth (tested with shibboleth sso)
- [x] Show login errors
- [x] Add new integration tests for token auth
- [x] Squash/arrange commits

BUGS:
- [x] setup is broken
- [x] user agent is always undefined, so the default is used
- [x] new user can not log in
- [x] ~~Changing own password kicks you out, hence the session token needs to be updated on password change~~ -> #24544

```
Exception: {"Exception":"OCP\\Files\\NotFoundException","Message":"","Code":0,"Trace":"#0 \/home\/christoph\/workspace\/owncloud\/apps\/files\/controller\/viewcontroller.php(118): OC_Helper::getStorageInfo('\/', false)\n#1 \/home\/christoph\/workspace\/owncloud\/apps\/files\/controller\/viewcontroller.php(180): OCA\\Files\\Controller\\ViewController->getStorageInfo()\n#2 [internal function]: OCA\\Files\\Controller\\ViewController->index('', '')\n#3 \/home\/christoph\/workspace\/owncloud\/lib\/private\/AppFramework\/Http\/Dispatcher.php(159): call_user_func_array(Array, Array)\n#4 \/home\/christoph\/workspace\/owncloud\/lib\/private\/AppFramework\/Http\/Dispatcher.php(89): OC\\AppFramework\\Http\\Dispatcher->executeController(Object(OCA\\Files\\Controller\\ViewController), 'index')\n#5 \/home\/christoph\/workspace\/owncloud\/lib\/private\/AppFramework\/App.php(110): OC\\AppFramework\\Http\\Dispatcher->dispatch(Object(OCA\\Files\\Controller\\ViewController), 'index')\n#6 \/home\/christoph\/workspace\/owncloud\/lib\/private\/AppFramework\/Routing\/RouteActionHandler.php(45): OC\\AppFramework\\App::main('OCA\\\\Files\\\\Contr...', 'index', Object(OC\\AppFramework\\DependencyInjection\\DIContainer), Array)\n#7 [internal function]: OC\\AppFramework\\Routing\\RouteActionHandler->__invoke(Array)\n#8 \/home\/christoph\/workspace\/owncloud\/lib\/private\/route\/router.php(280): call_user_func(Object(OC\\AppFramework\\Routing\\RouteActionHandler), Array)\n#9 \/home\/christoph\/workspace\/owncloud\/lib\/base.php(871): OC\\Route\\Router->match('\/apps\/files\/')\n#10 \/home\/christoph\/workspace\/owncloud\/index.php(39): OC::handleRequest()\n#11 {main}","File":"\/home\/christoph\/workspace\/owncloud\/lib\/private\/helper.php","Line":605}
```
- [x] ~~Not possible to add an account to the desktop client – I suspect the login url change to cause that. @LukasReschke what are we gonna do about that?~~ It works, I was testing with wrong credentials :see_no_evil: 
- [x] `"Unable to generate a URL for the named route \\\"login#showLoginForm\\\" as such route does not exist.`
- [x] Adjust column width
  `ORA-12899: value too large for column "AUTOTEST"."oc_authtoken"."token" (actual: 128, maximum: 100)`
- [x] username !== uid, https://github.com/owncloud/core/pull/24189#discussion_r60905030
- [x] Login with email doesnt work
## 

Next steps (follow-up PRs):
- GUI
- Clients should switch to token auth if the server supports it
- Bring back remember me, https://github.com/owncloud/core/pull/24189#discussion_r60743623

nickvergessen commented on apps/providerexample/appinfo/info.xml:
```diff
@@ -0,0 +1,16 @@
+	<default_enable/>
+	<types>
+		<!-- this is used to disable the feature of enabling an app for specific groups only because this would break this app -->
+		<filesystem/>
```
Remove, authentication also blocks disabling such apps

ChristophWurst replied:
This provider app example will be removed, it's just there for testing purposes

nickvergessen commented on apps/providerexample/appinfo/info.xml:
```diff
@@ -0,0 +1,16 @@
+	<description>Example provider for two factor auth</description>
+	<licence>AGPL</licence>
+	<author>Lukas Reschke</author>
+	<requiremin>8.2</requiremin>
```
old thing + wrong version, use dependency>owncloud

nickvergessen commented on apps/providerexample/appinfo/app.php:
```diff
@@ -0,0 +1,4 @@
+<?php
+
+$factory = \OC::$server->getTwoFactorAuthenticationFactory();
+$factory->registerProvider(new \OCA\providerexample\appinfo\StaticCodeProvider());
```
I don't think this app should live in core

nickvergessen commented on core/Auth/Controller/TwoFactorChallengeController.php:
```diff
@@ -0,0 +1,166 @@
+<?php
```
Should be in lib/private/Auth/Controller/TwoFactorChallengeController.php ?
With namespace OC\Auth\Controller ?

LukasReschke commented on lib/private/Authentication/Token/IProvider.php:
```diff
@@ -0,0 +1,32 @@
+
+	/**
+	 * @param string $token
+	 * @return boolean|string user UID if token is valid, false when invalid
```
I'm personally not a huge fan of APIs that return multiple types. Can we make it throw an exception when invalid?

ChristophWurst replied:
yes, using an exception sounds better

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -53,28 +64,36 @@
+	 * @param IProvider[] $tokenProvider
 	 */
-	public function __construct(\OCP\IUserManager $manager, \OCP\ISession $session) {
+	public function __construct(IUserManager $manager, ISession $session, $tokenProvider = []) {
```
`array $tokenProvider = []`?

LukasReschke commented on lib/public/iusersession.php:
```diff
@@ -49,6 +51,20 @@
+	 *
+	 * @since 9.1.0
+	 */
+	public function tryTokenLogin();
```
Personally I'd not add it to the interface in OCP. This would mean that we have to support this for 3 years which we might not want to.

Especially given the fact that I don't really see a lot of reasons for application developers to use these methods. This should be handled transparently by core.

LukasReschke commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -0,0 +1,51 @@
+
+	public function validateToken($token) {
+		$this->logger->debug('validating default token <' . $token . '>');
+		if ($token === 'rq778zYGX8') {
```
I know that this is far away from the final implementation but you might consider to use `hash_equals` on any comparison that might have security impact to avoid leaking timing information.

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -238,6 +257,71 @@ public function login($uid, $password) {
+	 *
+	 * @since 9.1.0
+	 */
+	public function tryTokenLogin() {
```
This looks already pretty good. There is one case that I'd however reconsider as well, so technically a "remember me" token could also just be a simple token provider here. So we might want to make this a bit more flexible since the remember me token will basically be a cookie.

So maybe let's add an additional method `verifyToken` that we can reuse for both places?

ChristophWurst replied:
Well tryTokenLogin already uses IRequest to get the header information. It could also be used to access the remember me cookie. Or do you prefer to keep those separate?

LukasReschke replied:
That's a very fair point. Let's also access the cookie in there then as well. 👍

LukasReschke commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -0,0 +1,51 @@
+			$this->logger->debug('token valid');
+			return 'user';
+		} else {
+			$this->logger->warning('invalid token');
```
Interesting. We might indeed want to look invalid token logins. For this we however need a mapping to avoid getting like a ton of warnings in case we have two providers. (such as the remember me token for example)

Also worth thinking about whether we want to throw some kind of "invalid auth" event at some point. But that's unrelated, just me thinking out loud :)

ChristophWurst replied:
So is there anything that needs to be changed in this PR or are those thought for future enhancements?

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -238,6 +257,71 @@ public function login($uid, $password) {
+		}
+	}
+
+	private function loginWithToken($uid, $token) {
```
$token is unused?

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -31,7 +33,16 @@
+use OC;
+use OC\Authentication\Token\IProvider;
 use OC\Hooks\Emitter;
+use OC\Session\Session;
```
That will fail some PHP versions. You can't `use` a class with the same name as the current class. Might need something like `use OC\Session\Session as WhateverSession`.

LukasReschke replied:
This is also what made Travis fail, we have a syntax check there :)

ChristophWurst committed:
token based auth
* Add InvalidTokenException
* add DefaultTokenMapper and use it to check if a auth token exists
* create new token for the browser session if none exists
hash stored token; save user agent
* encrypt login password when creating the token

LukasReschke commented on lib/private/util.php:
```diff
@@ -947,6 +947,44 @@ public static function checkDataDirectoryValidity($dataDirectory) {
+	 * @param array $errors
+	 * @param string[] $messages
+	 */
+	public static function displayLoginPage($errors = array(), $messages = []) {
```
Isn't that like, dead in master? :)

LukasReschke commented on lib/public/iusersession.php:
```diff
@@ -39,6 +39,7 @@
  * @since 6.0.0
  */
 interface IUserSession {
+
```
Unrequired

LukasReschke commented on core/Application.php:
```diff
@@ -1,6 +1,7 @@
 <?php
 /**
  * @author Bernhard Posselt <dev@bernhard-posselt.com>
+ * @author Christoph Wurst <christoph@winzerhof-wurst.at>
```
Let's agree either on owncloud.com or on winzerhof-wurst.at :)

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -238,6 +267,94 @@ public function login($uid, $password) {
+		return true;
+	}
+	
+	public function createSessionToken($uid, $password) {
```
PHPDoc :)

LukasReschke commented on core/Controller/LoginController.php:
```diff
@@ -150,11 +154,36 @@ public function showLoginForm($user,
+		if ($this->userManager->checkPassword($user, $password) === false) {
+			return new RedirectResponse($this->urlGenerator->linkToRoute('login#showLoginForm'));
+		}
+		$this->userSession->createSessionToken($user, $password);
```
I'd move that into it's own controller. Our web interface itself is unlikely to use that anytime soon.

So something like: A controller that takes user name + password (e.g. via POST?) and creates a new session token.

LukasReschke replied:
Hehe. And this will eat you as well. The login name doesn't has to be the UID :)

An LDAP user can login using their E-Mail address for example while their UID might be something totally different.

ChristophWurst replied:
What's the solutoin for this?

ChristophWurst replied:
@LukasReschke ^

LukasReschke commented on core/Controller/LoginController.php:
```diff
@@ -150,11 +154,36 @@ public function showLoginForm($user,
+			return new RedirectResponse($this->urlGenerator->linkToRoute('login#showLoginForm'));
+		}
+		$this->userSession->createSessionToken($user, $password);
+		if (isset($_REQUEST['redirect_url']) && OC_User::isLoggedIn()) {
```
IUserSession->isLoggedIn

LukasReschke replied:
While at it. Also adjust this to use `redirectUrl` as paramter? Or at least IRequest->getParam?

LukasReschke commented on core/Controller/LoginController.php:
```diff
@@ -150,11 +154,36 @@ public function showLoginForm($user,
+			// Deny the redirect if the URL contains a @
+			// This prevents unvalidated redirects like ?redirect_url=:user@domain.com
+			if (strpos($location, '@') === false) {
+				header('Location: ' . $location);
```
Simply a RedirectResponse

LukasReschke commented on lib/private/connector/sabre/auth.php:
```diff
@@ -0,0 +1,168 @@
+			$user = \OC::$server->getUserManager()->get($username);
+
+			if($user instanceof IUser &&
+				$user->isTwoFactorEnforced()) {
```
Leftover? :)

ChristophWurst replied:
The whole file is a leftover. This file lives in the dav app now

LukasReschke commented on lib/private/connector/sabre/auth.php:
```diff
@@ -0,0 +1,168 @@
+
+			if($user instanceof IUser &&
+				$user->isTwoFactorEnforced()) {
+				if(in_array($password, $user->getAuthenticationTokens(), true)) {
```
Where's the hashing happening here?

LukasReschke replied:
Good stuff btw. In some of the next steps I'd recommend to adjust the code to not use sessions at all. We probably want to use `\OC\Session\Memory` there and add some magic to not create any session cookie at all.

ChristophWurst replied:
> In some of the next steps

Is this something in the scope of this PR or can we do that in a follow-up PR? It's getting bigger and bigger and I'd like to keep this reviewable to lower the chance of introducing bugs

LukasReschke commented on core/templates/login.php:
```diff
@@ -9,7 +9,7 @@
 
 <!--[if IE 8]><style>input[type="checkbox"]{padding:0;}</style><![endif]-->
-<form method="post" name="login" action="<?php p(OC::$WEBROOT) ?>/">
+<form method="post" name="login" action="<?php p(OC::$WEBROOT) ?>/login">
```
Just remove the action URL as it goes against the same URL now.

LukasReschke commented on db_structure.xml:
```diff
@@ -1032,6 +1032,66 @@
+				<type>text</type>
+				<default></default>
+				<notnull>true</notnull>
+				<length>100</length>
```
Not a DB guru so I don't quite know what the "length" here has for implications but if this means: maximal 100 characters then this seems too fragile as the encryption logic will also add quite some characters (for the HMAC etc.)

ChristophWurst replied:
So how big should it be then? Who's the owncloud db guru? :)

LukasReschke replied:
Summoning the great guru @DeepDiver1975. 🙈 

I assume a clob with 4000 would be sufficient…?

LukasReschke commented on tests/lib/util.php:
```diff
@@ -278,64 +278,6 @@ public function dataProviderForTestIsSharingDisabledForUser() {
 	}
 
 	/**
-	 * Test default apps
```
Any reason that this got touched? 😉

ChristophWurst replied:
probably due to rebasing a six month old WIP-PR? :see_no_evil:

LukasReschke commented on tests/lib/util.php:
```diff
@@ -356,7 +298,6 @@ public function testGetDefaultPageUrlWithRedirectUrlRedirectBypassWithFrontContr
 		$this->assertSame('http://localhost'.\OC::$WEBROOT.'/apps/files/', OC_Util::getDefaultPageUrl());
 	}
 
-	/**
```
Same

LukasReschke commented on lib/public/iuser.php:
```diff
@@ -170,7 +170,7 @@ public function getAvatarImage($size);
 	 */
 	public function getCloudId();
-
+	
```
Unrequired

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -238,6 +267,94 @@ public function login($uid, $password) {
+	 * @param string $token
+	 * @return boolean
+	 */
+	private function validateToken(IRequest $request, $token) {
```
$request is missing in PHPDOc

ChristophWurst committed:
invalidate (delete) session token on logout
add 'last_activity' column to session tokens and delete old ones via a background job

ChristophWurst commented:
Weir error when running ``cron.php``:
```
{"reqId":"9BIQ9E8gnuskJVSjcO09","remoteAddr":"","app":"core","message":"Error while running background job (class: OC\\Authentication\\Token\\DefaultTokenCleanupJob, arguments: ): {\"Exception\":\"OCP\\\\AppFramework\\\\QueryException\",\"Message\":\"Could not resolve OCP\\\\IDb! Class can not be instantiated\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(104): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->resolve('OCP\\\\\\\\IDb')\\n#1 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/servercontainer.php(87): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->query('OCP\\\\\\\\IDb')\\n#2 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(64): OC\\\\ServerContainer->query('OCP\\\\\\\\IDb')\\n#3 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(83): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->buildClass(Object(ReflectionClass))\\n#4 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(104): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->resolve('OC\\\\\\\\Authenticati...')\\n#5 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/servercontainer.php(87): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->query('OC\\\\\\\\Authenticati...')\\n#6 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(64): OC\\\\ServerContainer->query('OC\\\\\\\\Authenticati...')\\n#7 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(83): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->buildClass(Object(ReflectionClass))\\n#8 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(104): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->resolve('OC\\\\\\\\Authenticati...')\\n#9 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/servercontainer.php(87): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->query('OC\\\\\\\\Authenticati...')\\n#10 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/Authentication\\\/Token\\\/DefaultTokenCleanupJob.php(33): OC\\\\ServerContainer->query('OC\\\\\\\\Authenticati...')\\n#11 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/backgroundjob\\\/job.php(52): OC\\\\Authentication\\\\Token\\\\DefaultTokenCleanupJob->run(NULL)\\n#12 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/cron.php(145): OC\\\\BackgroundJob\\\\Job->execute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#13 {main}\",\"File\":\"\\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php\",\"Line\":85}","level":3,"time":"2016-04-25T15:45:23+00:00","method":"--","url":"--","user":"--"}
```
@DeepDiver1975 didn't we have that exact same error in the mail app some weeks back and you magically solved it? As far as I remember there is a problem with the ``IDb``/``IDbConnection`` interfaces app framework mappers need. Do you know what the fix was?

ChristophWurst committed:
Check if session token is valid and log user out if the check fails
* Update last_activity timestamp of the session token
* Check user backend credentials once in 5 minutes

ChristophWurst committed:
Add controller to generate client tokens

ChristophWurst committed:
Add index on 'last_activity'

add token type column and delete only temporary tokens in the background job

debounce token updates; fix wrong class import

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -143,27 +170,61 @@ public function setUser($user) {
+		}
+
+		// Session is valid, so the token can be refreshed
+		$this->tokenProvider->updateToken($token);
```
Are we sure that's gonna perform fine? I can imagine this can hammer the database quite a lot? Maybe we should only do this action like 1 a minute?

ChristophWurst replied:
Let's debounce that too

ChristophWurst commented on lib/private/user/session.php:
```diff
@@ -143,27 +169,66 @@ public function setUser($user) {
+		try {
+			$token = $this->tokenProvider->getToken($sessionId);
+		} catch (InvalidTokenException $ex) {
+			// Session was inalidated
```
:see_no_evil:

ChristophWurst committed:
Add token auth for OCS APIs

ChristophWurst committed:
Add fallback to allow user:token basic auth

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -278,6 +278,13 @@ public function login($uid, $password) {
 		$user = $this->manager->checkPassword($uid, $password);
+		if ($user === false) {
+			// Password auth failed, maybe it's a token
+			$request = \OC::$server->getRequest();
```
Inject? 😉

ChristophWurst replied:
Not possible due to cyclic dependencies, see ``tryTokenLogin`` :-(

ChristophWurst replied:
Also, passing in the IRequest object is also not an option, as that would require to change the public IUserSession interface

LukasReschke replied:
Fair enough.

rullzer replied:
I don't get this... why would this need changes to the interface?

ChristophWurst replied:
I cannot inject IRequest via the session constructor due to cyclic dependencies. The alternative, passing the IRequest object as parameter of the login method would change the interface. Hence that's also not possible without potentially breaking apps extending the interface.

ChristophWurst committed:
increase version number to trigger db migration

ChristophWurst committed:
fix setup

ChristophWurst commented on lib/private/user/session.php:
```diff
@@ -217,13 +285,27 @@ public function login($uid, $password) {
+						$user = $this->getUser()->getUID();
+						\OC_Util::setupFS($user);
+						//trigger creation of user home and /files folder
+						\OC::$server->getUserFolder($user);
```
TODO: inject those instances if possible

ChristophWurst committed:
fix LoginController unit tests

ChristophWurst committed:
Fix existing tests

ChristophWurst committed:
increase token column width

add some range to time() assertions

ChristophWurst committed:
try apache auth too

ChristophWurst committed:
pass in $request on OCS api

ChristophWurst committed:
add unit tests for all new classes

ChristophWurst committed:
fix mock builder for old phpunit versions

ChristophWurst commented:
- [ ] BUG: sync client sessions are killed because there is no session token entry in the DB

I tried to implement this as generic as possible, so no special hacks are needed for various endpoints. ``IUserSession::isLoggedIn()`` uses ``IUserSession::getUser()`` to determine if a user is logged in, based on a session value (``user_id``). Additionally, this method also checks if a session token entry exists in the DB, and logs the user out automatically if none can be found. However, also the DAV endpoint uses ``IUserSession::isLoggedIn()``, but there we never generate a session token for the client because there is no explicit *client login*, the login happens on each request if necessary.

@LukasReschke any idea how we could solve this? Should we bypass that check on DAV? Should we create a session token entry for sync clients?

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -237,6 +310,136 @@ public function login($uid, $password) {
+	 * Tries to login the user with HTTP Basic Authentication
+	 *
+	 * @param IRequest $request
+	 * @return boolean if the login was successful
```
`$this->login` claims also to return `null`. From looking at the PHPDoc this seems to be however wrong, so maybe update it's PHPDoc?

LukasReschke commented on core/Controller/TokenController.php:
```diff
@@ -0,0 +1,87 @@
+	 *
+	 * @param string $user
+	 * @param string $password
+	 */
```
![2016-05-02_16-23-28](https://cloud.githubusercontent.com/assets/878997/14956863/3516e220-1082-11e6-8ae8-84490f33ccc4.png)

LukasReschke commented on core/Controller/TokenController.php:
```diff
@@ -0,0 +1,87 @@
+	 * @param string $user
+	 * @param string $password
+	 */
+	public function generateToken($user, $password, $name = 'unknown client') {
```
![2016-05-02_16-23-44](https://cloud.githubusercontent.com/assets/878997/14956870/3fd38be6-1082-11e6-85aa-1b1d579def6c.png)

LukasReschke commented on core/Controller/TokenController.php:
```diff
@@ -0,0 +1,87 @@
+	 */
+	public function generateToken($user, $password, $name = 'unknown client') {
+		if (is_null($user) || is_null($password)) {
+			$response = new Response([]);
```
Does response really have a constructor that takes an array?

LukasReschke commented on core/Controller/TokenController.php:
```diff
@@ -0,0 +1,87 @@
+			return $response;
+		}
+		if ($this->userManager->checkPassword($user, $password) === false) {
+			$response = new Response([]);
```
Does response really have a constructor that takes an array?

LukasReschke commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -0,0 +1,187 @@
+	 * @param string $token
+	 * @param string $uid
+	 * @param string $password
+	 * @apram int $type token type
```
Typo :)

LukasReschke commented on lib/private/Authentication/Token/DefaultToken.php:
```diff
@@ -0,0 +1,67 @@
+
+use OCP\AppFramework\Db\Entity;
+
+class DefaultToken extends Entity implements IToken {
```
Please annotate the magic setters and getters: https://manual.phpdoc.org/HTMLSmartyConverter/PHP/phpDocumentor/tutorial_tags.property.pkg.html

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -237,6 +310,136 @@ public function login($uid, $password) {
+	 * @param IRequest $request
+	 * @return boolean if the login was successful
+	 */
+	public function tryBasicAuthLogin(IRequest $request) {
```
What's `IRequest` here for?

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -237,6 +310,136 @@ public function login($uid, $password) {
+	private function updateToken(IProvider $provider, IToken $token) {
+		// To save unnecessary DB queries, this is only done once a minute
+		$lastTokenUpdate = $this->session->get('last_token_update') ? : 0;
+		if ($lastTokenUpdate < (time() - 60)) {
```
Use the ITimeFactory

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -237,6 +310,136 @@ public function login($uid, $password) {
+		$lastTokenUpdate = $this->session->get('last_token_update') ? : 0;
+		if ($lastTokenUpdate < (time() - 60)) {
+			$provider->updateToken($token);
+			$this->session->set('last_token_update', time());
```
Use the ITimeFactory

LukasReschke commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -0,0 +1,187 @@
+	 * @return string
+	 */
+	private function hashToken($token) {
+		return hash('sha512', $token);
```
I'd also propose to postfix this with the application secret from config.php

LukasReschke commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -0,0 +1,187 @@
+	 * Invalidate (delete) old session tokens
+	 */
+	public function invalidateOldTokens() {
+		$olderThan = time() - (int) $this->config->getSystemValue('session_lifetime', 60 * 60 * 24);
```
ITimeFactory

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -143,25 +177,56 @@ public function setUser($user) {
+		// Check whether login credentials are still valid
+		// This check is performed each 5 minutes
+		$lastCheck = $this->session->get('last_login_check') ? : 0;
+		if ($lastCheck < (time() - 60 * 5)) {
```
ITimeFactory

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -143,25 +177,56 @@ public function setUser($user) {
+	}
+
+	protected function validateSession(IUser $user) {
+		// TODO: use ISession::getId(), https://github.com/owncloud/core/pull/24229
```
Merged now

ChristophWurst committed:
PHPDoc and other minor fixes

ChristophWurst committed:
add auth integration tests

ChristophWurst committed:
catch possible SessionNotAvailableExceptions

ChristophWurst committed:
show login error

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -55,22 +69,55 @@
+
+	/**
+	 * @var User $activeUser
 	protected $activeUser;
```
Syntax error? 🙈

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -55,22 +69,55 @@
+	 * @param IProvider[] $tokenProviders
 	 */
-	public function __construct(IUserManager $manager, ISession $session) {
+	public function __construct(IUserManager $manager, ISession $session, ITimeFactory $timeFacory, $tokenProvider,
```
Update PHPDoc

rullzer replied:
I realize this was already the case. But this class is not injected with an IUserManager. It is expecting a Manager. Else a lot of stuff downwards also breaks...

LukasReschke commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -0,0 +1,193 @@
+	 * @param IConfig $config
+	 * @param ILogger $logger
+	 */
+	public function __construct(DefaultTokenMapper $mapper, ICrypto $crypto, IConfig $config, ILogger $logger, ITimeFactory $time) {
```
PHPDoc is not updated

rullzer commented on core/Controller/LoginController.php:
```diff
@@ -150,11 +153,49 @@ public function showLoginForm($user,
+				$loginResult = $this->userManager->checkPassword($users[0]->getUID(), $password);
+			}
+		}
+		if (!$loginResult) {
```
Merge with above?

ChristophWurst replied:
How? Move the if into the if above?

LukasReschke commented on build/integration/features/auth.feature:
```diff
@@ -0,0 +1,78 @@
+	Scenario: using OCS with browser session
+		Given a new browser session is started
+		When requesting "/ocs/v1.php/apps/files_sharing/api/v1/remote_shares" with "GET" using browser session
+		Then the OCS status code should be "100"
```
🚀 🙌 🎉 🎂  - Yay. Tests 😉

rullzer replied:
:clap:

rullzer commented on core/Controller/TokenController.php:
```diff
@@ -0,0 +1,89 @@
+		}
+		$token = $this->secureRandom->generate(128);
+		$this->tokenProvider->generateToken($token, $user, $password, $name, IToken::PERMANENT_TOKEN);
+		return [
```
Better create a JSONResponse here to return...

LukasReschke replied:
Personally I'm also happy with an array, but then the PHPDoc should be fixed :)

We have some automagical converters, if you return an array it will be returned in the format you converted. Planned was that it also would have been able to return XML. But until now we only have JSON 😉

rullzer replied:
Well fine with me but then indeed update the docs ;)

LukasReschke commented on lib/private/Updater.php:
```diff
@@ -216,6 +216,7 @@ private function doUpgrade($currentVersion, $installedVersion) {
 		try {
 			Setup::updateHtaccess();
 			Setup::protectDataDirectory();
+			Setup::installBackgroundJobs();
```
Add TODO to replace this with https://github.com/owncloud/core/pull/24378 once merged?

rullzer commented on lib/private/Authentication/Exceptions/InvalidTokenException.php:
```diff
@@ -0,0 +1,29 @@
+
+use Exception;
+
+class InvalidTokenException extends Exception {
```
Maybe extend the hintexception so you can add some error text if this ever surfaces?

rullzer commented on lib/private/Authentication/Token/DefaultToken.php:
```diff
@@ -0,0 +1,81 @@
+		return $this->id;
+	}
+
+	public function getUid() {
```
Maybe for consitency use `UID`. We have that in other parts of the code as well

rullzer commented on lib/private/Authentication/Token/DefaultTokenMapper.php:
```diff
@@ -0,0 +1,77 @@
+	 * @param string $token
+	 */
+	public function invalidate($token) {
+		$sql = 'DELETE FROM `' . $this->getTableName() . '` '
```
Please us the query builder!

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -139,25 +186,60 @@ public function setUser($user) {
+		try {
+			$sessionId = $this->session->getId();
+		} catch (SessionNotAvailableException $ex) {
+			return;
```
The return values confuse me. So this function can either return void or a bool but none of them seem to get checked anyways. So maybe just use void? :)

rullzer commented on lib/private/Authentication/Token/DefaultTokenMapper.php:
```diff
@@ -0,0 +1,77 @@
+	/**
+	 * Invalidate (delete) a given token
+	 *
+	 * @param string $token
```
Missing `@return`

rullzer commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -0,0 +1,193 @@
+	 * @param string $password
+	 * @param int $type token type
+	 * @return DefaultToken
+	 */
```
Missing phpdoc for `$name`

rullzer commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -0,0 +1,193 @@
+	/**
+	 * Update token activity timestamp
+	 *
+	 * @param DefaultToken $token
```
missing `@throws`

rullzer replied:
Also this `@param` doesn't match

rullzer commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -0,0 +1,193 @@
+
+	/**
+	 * @param string $token
+	 * @throws InvalidTokenException
```
missing `@return`

rullzer commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -0,0 +1,193 @@
+
+	/**
+	 * @param DefaultToken $savedToken
+	 * @param string $token session token
```
Missing `@return`

rullzer commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -0,0 +1,193 @@
+	 * @param DefaultTokenMapper $mapper
+	 * @param ICrypto $crypto
+	 * @param IConfig $config
+	 * @param ILogger $logger
```
Missing `@param` form ITimeFactory

LukasReschke commented on core/Controller/LoginController.php:
```diff
@@ -96,18 +100,17 @@ public function logout() {
 		}
 
 		$parameters = array();
+		$id = $this->session->getId();
```
$id is never used?

LukasReschke commented on core/Controller/LoginController.php:
```diff
@@ -150,11 +153,49 @@ public function showLoginForm($user,
+			}
+		}
+		if (!$loginResult) {
+			$id = $this->session->getId();
```
$id is never used

LukasReschke commented on lib/private/legacy/api.php:
```diff
@@ -353,35 +353,25 @@ private static function loginUser() {
+		try {
+			$loginSuccess = $userSession->tryTokenLogin($request);
+			if (!$loginSuccess) {
+				$loginSuccess = $userSession->tryBasicAuthLogin();
```
Missing $request here

ChristophWurst replied:
I don't get it. Where is it missing exactly?

LukasReschke replied:
In `$userSession->tryBasicAuthLogin`, see https://github.com/owncloud/core/pull/24189/files#diff-b1f73233891e998033d16255fc6d5db8R350 for the arguments

rullzer commented on lib/private/Authentication/Token/IProvider.php:
```diff
@@ -0,0 +1,42 @@
+	/**
+	 * Update token activity timestamp
+	 *
+	 * @param DefaultToken $token
```
Does not match signature

LukasReschke commented on core/Controller/LoginController.php:
```diff
@@ -96,18 +100,17 @@ public function logout() {
 	 *
 	 * @return TemplateResponse
```
Requires adjusting as it also can be a RedirectResponse

rullzer commented on lib/private/Files/Filesystem.php:
```diff
@@ -404,7 +404,7 @@ public static function initMountPoints($user = '') {
 		if (is_null($userObject)) {
 			\OCP\Util::writeLog('files', ' Backends provided no user object for ' . $user, \OCP\Util::ERROR);
-			throw new \OC\User\NoUserException('Backends provided no user object for ' . $user);
+			throw new \OC\User\NoUserException('Backend provided no user object for ' . $user);
```
Intentional change?

ChristophWurst replied:
Hm, I don't remember :see_no_evil:

rullzer commented on lib/private/user/session.php:
```diff
@@ -55,22 +69,55 @@
+	private $tokenProviders;
+
+	/**
+	 * @var User $activeUser
```
missng `*/`

rullzer commented on lib/private/user/session.php:
```diff
@@ -55,22 +69,55 @@
 	/**
 	 * @param IUserManager $manager
 	 * @param ISession $session
+	 * @param IProvider[] $tokenProviders
```
phpdoc incomplete

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -241,6 +331,143 @@ public function login($uid, $password) {
+	}
+
+	private function loginWithToken($uid) {
+		//$this->manager->emit('\OC\User', 'preTokenLogin', array($uid));
```
Add TODO in front so static source code analyzers can find this :)

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -241,6 +331,143 @@ public function login($uid, $password) {
+
+		//login
+		$this->setUser($user);
+		//$this->manager->emit('\OC\User', 'postTokenLogin', array($user));
```
Add TODO in front so static source code analyzers can find this :)

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -293,30 +528,31 @@ public function logout() {
+		$expires = time() + OC::$server->getConfig()->getSystemValue('remember_login_cookie_lifetime', 60 * 60 * 24 * 15);
+		setcookie("oc_username", $username, $expires, OC::$WEBROOT, '', $secureCookie, true);
+		setcookie("oc_token", $token, $expires, OC::$WEBROOT, '', $secureCookie, true);
+		setcookie("oc_remember_login", "1", $expires, OC::$WEBROOT, '', $secureCookie, true);
```
While touching this you may as well switch all to `'` instead of `"` here :)

rullzer commented on lib/private/user/session.php:
```diff
@@ -241,6 +331,143 @@ public function login($uid, $password) {
+	protected function prepareUserLogin() {
+		// TODO: mock/inject/use non-static
+		// Refresh the token
+		\OC::$server->getCsrfTokenManager()->refreshToken();
```
Inject... as the todo says :)

ChristophWurst replied:
Cyclic dependency ``Maximum function nesting level of '256' reached, aborting!`` :-/

UserSession -> CsrfTokenManager -> UserSession :boom:

rullzer replied:
Ah warg. seems like we need to  introduce a level of laziness somewhere...

LukasReschke commented:
There is a bug in case the database has not yet been updated but somebody with a session opens ownCloud.

1. Have master
2. Login
3. Change to this branch
4. Refresh

```
Type: Doctrine\DBAL\Exception\TableNotFoundException
Code: 0
Message: An exception occurred while executing 'SELECT "id", "uid", "password", "name", "token", "last_activity" FROM "oc_authtoken" WHERE "token" = ?': SQLSTATE[HY000]: General error: 1 no such table: oc_authtoken
File: /Users/lukasreschke/Documents/Programming/oc/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractSQLiteDriver.php
Line: 58

Trace
#0 /Users/lukasreschke/Documents/Programming/oc/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php(116): Doctrine\DBAL\Driver\AbstractSQLiteDriver->convertException('An exception oc...', Object(Doctrine\DBAL\Driver\PDOException))
#1 /Users/lukasreschke/Documents/Programming/oc/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(785): Doctrine\DBAL\DBALException::driverExceptionDuringQuery(Object(Doctrine\DBAL\Driver\PDOSqlite\Driver), Object(Doctrine\DBAL\Driver\PDOException), 'SELECT "id", "u...')
#2 /Users/lukasreschke/Documents/Programming/oc/lib/private/DB/Connection.php(162): Doctrine\DBAL\Connection->prepare('SELECT "id", "u...')
#3 /Users/lukasreschke/Documents/Programming/oc/lib/public/AppFramework/Db/Mapper.php(234): OC\DB\Connection->prepare('SELECT `id`, `u...', NULL, NULL)
#4 /Users/lukasreschke/Documents/Programming/oc/lib/public/AppFramework/Db/Mapper.php(280): OCP\AppFramework\Db\Mapper->execute('SELECT `id`, `u...', Array, NULL, NULL)
#5 /Users/lukasreschke/Documents/Programming/oc/lib/public/AppFramework/Db/Mapper.php(372): OCP\AppFramework\Db\Mapper->findOneQuery('SELECT `id`, `u...', Array, NULL, NULL)
#6 /Users/lukasreschke/Documents/Programming/oc/lib/private/Authentication/Token/DefaultTokenMapper.php(74): OCP\AppFramework\Db\Mapper->findEntity('SELECT `id`, `u...', Array)
#7 /Users/lukasreschke/Documents/Programming/oc/lib/private/Authentication/Token/DefaultTokenProvider.php(107): OC\Authentication\Token\DefaultTokenMapper->getToken('19a02eeee04cc33...')
#8 /Users/lukasreschke/Documents/Programming/oc/lib/private/User/Session.php(218): OC\Authentication\Token\DefaultTokenProvider->getToken('qbsftenpk9694tn...')
#9 /Users/lukasreschke/Documents/Programming/oc/lib/private/User/Session.php(206): OC\User\Session->validateSession(Object(OC\User\User))
#10 /Users/lukasreschke/Documents/Programming/oc/lib/private/L10N/Factory.php(131): OC\User\Session->getUser()
#11 /Users/lukasreschke/Documents/Programming/oc/lib/private/L10N/Factory.php(99): OC\L10N\Factory->findLanguage('core')
#12 /Users/lukasreschke/Documents/Programming/oc/lib/private/Server.php(895): OC\L10N\Factory->get('core', NULL)
#13 /Users/lukasreschke/Documents/Programming/oc/lib/private/Server.php(123): OC\Server->getL10N('core')
#14 /Users/lukasreschke/Documents/Programming/oc/3rdparty/pimple/pimple/src/Pimple/Container.php(113): OC\Server->OC\{closure}(Object(OC\Server))
#15 /Users/lukasreschke/Documents/Programming/oc/lib/private/AppFramework/Utility/SimpleContainer.php(102): Pimple\Container->offsetGet('EncryptionManag...')
#16 /Users/lukasreschke/Documents/Programming/oc/lib/private/ServerContainer.php(87): OC\AppFramework\Utility\SimpleContainer->query('EncryptionManag...')
#17 /Users/lukasreschke/Documents/Programming/oc/lib/private/Server.php(681): OC\ServerContainer->query('EncryptionManag...')
#18 /Users/lukasreschke/Documents/Programming/oc/lib/base.php(726): OC\Server->getEncryptionManager()
#19 /Users/lukasreschke/Documents/Programming/oc/lib/base.php(645): OC::registerEncryptionWrapper()
#20 /Users/lukasreschke/Documents/Programming/oc/lib/base.php(947): OC::init()
#21 /Users/lukasreschke/Documents/Programming/oc/index.php(37): require_once('/Users/lukasres...')
#22 {main}
```

rullzer commented:
Already starting to look very nice! Kudos!

Could you fix the syntax error so CI can at least run?
I'll have a more in depth look then.

LukasReschke commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -0,0 +1,193 @@
+	 */
+	private function decryptPassword($password, $token) {
+		$secret = $this->config->getSystemValue('secret');
+		return $this->crypto->decrypt($password, $token . $secret);
```
This can also throw an exception if the decryption somehow failed (e.g. somehow tampered with the DB).

LukasReschke replied:
In my opinion when an exception is thrown the session should be invalidated (preferred) or an empty password returned (not really my preference but should do as well)

LukasReschke commented:
From @ChristophWurst's https://github.com/owncloud/core/pull/24189#issuecomment-216246678: 
> - [ ] BUG: sync client sessions are killed because there is no session token entry in the DB
> 
> I tried to implement this as generic as possible, so no special hacks are needed for various endpoints. ``IUserSession::isLoggedIn()`` uses ``IUserSession::getUser()`` to determine if a user is logged in, based on a session value (``user_id``). Additionally, this method also checks if a session token entry exists in the DB, and logs the user out automatically if none can be found. However, also the DAV endpoint uses ``IUserSession::isLoggedIn()``, but there we never generate a session token for the client because there is no explicit *client login*, the login happens on each request if necessary.
> 
> @LukasReschke any idea how we could solve this? Should we bypass that check on DAV? Should we create a session token entry for sync clients?

Personally I'd go with the minimum viable product way here:

Step 1: Bypass that check on DAV (in this PR)
Step 2: Add logic to detect if the DAV client supports sessions (in another PR). We don't want to spam this completely with DAV clients that can't do sessions.
Step 3: If DAV client supports sessions verify them as well.

I can take care of step 2 and 3 in another PR. But first let's get this as far as possible.

LukasReschke commented:
🚀 😄 🦄 🦁  🎉

ChristophWurst committed:
fix PHPDoc and other minor issues

ChristophWurst commented on tests/lib/authentication/token/defaulttokenmappertest.php:
```diff
@@ -0,0 +1,131 @@
+			->from('authtoken')
+			->execute()
+			->fetch();
+		print_r($result);
```
- [ ] TODO: remove

ChristophWurst commented:
> Could you fix the syntax error so CI can at least run?
> I'll have a more in depth look then.

@rullzer done. http://classicprogrammerpaintings.com/post/142841291544/waiting-for-the-build-also-known-as-scala

ChristophWurst commented:
- [ ] BUG: ``PHP Fatal error:  Uncaught TypeError: Argument 1 passed to OC\Authentication\Token\DefaultTokenMapper::__construct() must implement interface OCP\IDBConnection, none given, called in phar:///usr/local/bin/phpunit/phpunit-mock-objects/Framework/MockObject/Generator.php on line 254 and defined in /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/lib/private/Authentication/Token/DefaultTokenMapper.php:31``

ChristophWurst committed:
use the query builder instead of raw sql statements

ChristophWurst committed:
delete the token in case an exception is thrown when decrypting the password

ChristophWurst committed:
fix login with email

ChristophWurst commented:
> There is a bug in case the database has not yet been updated but somebody with a session opens ownCloud. (works if one deletes the cookies locally)

@LukasReschke  I cannot reproduce this.

What I tried:

1. change to master
2. empty authtoken table
3. make sure user is still logged in
4. switch to pluggable auth branch
5. reload page

Then I see the update page, click update and the update is run successfully. Afterwards, I'm redirected to the login page, as no session token can be found and the user is logged out. So no issues detected, everything works as expected.

ChristophWurst committed:
create session token for DAV clients (sync clients)

LukasReschke commented on lib/private/user/session.php:
```diff
@@ -241,6 +336,141 @@ public function login($uid, $password) {
+	/**
+	 * Tries to login the user with auth token header
+	 *
+	 * @todo check remember me cookie
```
Missing return value :)

ChristophWurst committed:
use the UID for creating the session token, not the login name

ChristophWurst commented:
Failing build jobs are unrelated

rullzer commented:
Looks good! I'll run the itnergration tests locally to check if things burn down...

rullzer commented:
All pass! I say lets get this in! :+1:

LukasReschke commented:
👍

skjnldsv commented:
Impressive work @ChristophWurst !! :D

DeepDiver1975 commented:
rebased to get the latest status of jenkins ...

ChristophWurst commented:
> This branch has conflicts that must be resolved

@DeepDiver1975 could you please rebase again?

DeepDiver1975 commented:
> @DeepDiver1975 could you please rebase again?

will do once the first ci statuses pop in ....

guruz commented:
> Clients should switch to token auth if the server supports it

@dragotin @danimo How does this relate to the discussions you had yesterday?
